# Data Preprocessing
In this first section, the dataset, in its original form, is loaded and analyzed to get an insight on it's features domains, values and meaning; successively, the data is preprocessed in order to create a cleaner csv file as well as to generate additional features that will (hopefully) improve the predictive maintenance model performance.

## Libraries imports
The library Pandas will be extensively used throughout the entire notebook, since it is the easiest and most powerful environment to work with in case of datasets managing

In [1]:
import pandas as pd
import numpy as np

## Data Loading
The train, test and truth datasets are loaded from the memory and modified into a more organized structure.

Loading the train data:

In [2]:
df_train_raw = pd.read_csv('data/PM_train.txt', sep = ' ', header=None)
df_train_raw

0    1       2       3      4       5       6        7        8   \
0        1    1 -0.0007 -0.0004  100.0  518.67  641.82  1589.70  1400.60   
1        1    2  0.0019 -0.0003  100.0  518.67  642.15  1591.82  1403.14   
2        1    3 -0.0043  0.0003  100.0  518.67  642.35  1587.99  1404.20   
3        1    4  0.0007  0.0000  100.0  518.67  642.35  1582.79  1401.87   
4        1    5 -0.0019 -0.0002  100.0  518.67  642.37  1582.85  1406.22   
...    ...  ...     ...     ...    ...     ...     ...      ...      ...   
20626  100  196 -0.0004 -0.0003  100.0  518.67  643.49  1597.98  1428.63   
20627  100  197 -0.0016 -0.0005  100.0  518.67  643.54  1604.50  1433.58   
20628  100  198  0.0004  0.0000  100.0  518.67  643.42  1602.46  1428.18   
20629  100  199 -0.0011  0.0003  100.0  518.67  643.23  1605.26  1426.53   
20630  100  200 -0.0032 -0.0005  100.0  518.67  643.85  1600.38  1432.14   

          9   ...       18      19    20   21    22     23     24       25  \
0      14.62  ...  8138.62  8.4195  0.03  392  2388  100.0  39.06  23.4190   
1      14.62  ...  8131.49  8.4318  0.03  392  2388  100.0  39.00  23.4236   
2      14.62  ...  8133.23  8.4178  0.03  390  2388  100.0  38.95  23.3442   
3      14.62  ...  8133.83  8.3682  0.03  392  2388  100.0  38.88  23.3739   
4      14.62  ...  8133.80  8.4294  0.03  393  2388  100.0  38.90  23.4044   
...      ...  ...      ...     ...   ...  ...   ...    ...    ...      ...   
20626  14.62  ...  8137.60  8.4956  0.03  397  2388  100.0  38.49  22.9735   
20627  14.62  ...  8136.50  8.5139  0.03  395  2388  100.0  38.30  23.1594   
20628  14.62  ...  8141.05  8.5646  0.03  398  2388  100.0  38.44  22.9333   
20629  14.62  ...  8139.29  8.5389  0.03  395  2388  100.0  38.29  23.0640   
20630  14.62  ...  8137.33  8.5036  0.03  396  2388  100.0  38.37  23.0522   

       26  27  
0     NaN NaN  
1     NaN NaN  
2     NaN NaN  
3     NaN NaN  
4     NaN NaN  
...    ..  ..  
20626 NaN NaN  
20627 NaN NaN  
20628 NaN NaN  
20629 NaN NaN  
20630 NaN NaN  

[20631 rows x 28 columns]

We see that the last two columns are unused, we proceed to drop them and assign names to all columns according to their physical meaning

In [3]:
df_train_raw.drop([26,27], axis=1, inplace=True)
col_names = ['id','cycle','setting1','setting2','setting3','s1','s2','s3','s4','s5','s6','s7','s8','s9','s10','s11','s12','s13','s14','s15','s16','s17','s18','s19','s20','s21']
df_train_raw.columns = col_names

### Train Dataset description
The features in the train (and test) dataset represent paramenter values registered by each airplane engine at every cycle, they are arranged as follows:
* **id**: the engine identification number, in this case it is an integer varying from 1 to 100;
* **cycle**: the cycle during which the values were registered, it acts as the timestamp variable;
* **setting1** to **setting3**: engine operational settings;
* **s1** to **s21**: the 21 different sensors that capture engine paramenters during each cycle, all together represent the engine state.

In [4]:
df_train_raw.head()

id  cycle  setting1  setting2  setting3      s1      s2       s3       s4  \
0   1      1   -0.0007   -0.0004     100.0  518.67  641.82  1589.70  1400.60   
1   1      2    0.0019   -0.0003     100.0  518.67  642.15  1591.82  1403.14   
2   1      3   -0.0043    0.0003     100.0  518.67  642.35  1587.99  1404.20   
3   1      4    0.0007    0.0000     100.0  518.67  642.35  1582.79  1401.87   
4   1      5   -0.0019   -0.0002     100.0  518.67  642.37  1582.85  1406.22   

      s5  ...     s12      s13      s14     s15   s16  s17   s18    s19  \
0  14.62  ...  521.66  2388.02  8138.62  8.4195  0.03  392  2388  100.0   
1  14.62  ...  522.28  2388.07  8131.49  8.4318  0.03  392  2388  100.0   
2  14.62  ...  522.42  2388.03  8133.23  8.4178  0.03  390  2388  100.0   
3  14.62  ...  522.86  2388.08  8133.83  8.3682  0.03  392  2388  100.0   
4  14.62  ...  522.19  2388.04  8133.80  8.4294  0.03  393  2388  100.0   

     s20      s21  
0  39.06  23.4190  
1  39.00  23.4236  
2  38.95  23.3442  
3  38.88  23.3739  
4  38.90  23.4044  

[5 rows x 26 columns]

Now we check if the dataset contains any missing value, something that would create some problems in the modeling phase:

In [5]:
df_train_raw.isnull().sum()

id          0
cycle       0
setting1    0
setting2    0
setting3    0
s1          0
s2          0
s3          0
s4          0
s5          0
s6          0
s7          0
s8          0
s9          0
s10         0
s11         0
s12         0
s13         0
s14         0
s15         0
s16         0
s17         0
s18         0
s19         0
s20         0
s21         0
dtype: int64

Luckily the dataset provider was careful to remove any (if present) missing (`NaN`) value, so we can continue without having to deal with the realated problems.
A good approach to any new dataset that we encounter is to watch the information given by the `describe` function from <span style="color:#90ee90">**Pandas**</span>:

In [6]:
df_train_raw.describe()

id         cycle      setting1      setting2  setting3  \
count  20631.000000  20631.000000  20631.000000  20631.000000   20631.0   
mean      51.506568    108.807862     -0.000009      0.000002     100.0   
std       29.227633     68.880990      0.002187      0.000293       0.0   
min        1.000000      1.000000     -0.008700     -0.000600     100.0   
25%       26.000000     52.000000     -0.001500     -0.000200     100.0   
50%       52.000000    104.000000      0.000000      0.000000     100.0   
75%       77.000000    156.000000      0.001500      0.000300     100.0   
max      100.000000    362.000000      0.008700      0.000600     100.0   

             s1            s2            s3            s4            s5  ...  \
count  20631.00  20631.000000  20631.000000  20631.000000  2.063100e+04  ...   
mean     518.67    642.680934   1590.523119   1408.933782  1.462000e+01  ...   
std        0.00      0.500053      6.131150      9.000605  1.776400e-15  ...   
min      518.67    641.210000   1571.040000   1382.250000  1.462000e+01  ...   
25%      518.67    642.325000   1586.260000   1402.360000  1.462000e+01  ...   
50%      518.67    642.640000   1590.100000   1408.040000  1.462000e+01  ...   
75%      518.67    643.000000   1594.380000   1414.555000  1.462000e+01  ...   
max      518.67    644.530000   1616.910000   1441.490000  1.462000e+01  ...   

                s12           s13           s14           s15           s16  \
count  20631.000000  20631.000000  20631.000000  20631.000000  2.063100e+04   
mean     521.413470   2388.096152   8143.752722      8.442146  3.000000e-02   
std        0.737553      0.071919     19.076176      0.037505  1.387812e-17   
min      518.690000   2387.880000   8099.940000      8.324900  3.000000e-02   
25%      520.960000   2388.040000   8133.245000      8.414900  3.000000e-02   
50%      521.480000   2388.090000   8140.540000      8.438900  3.000000e-02   
75%      521.950000   2388.140000   8148.310000      8.465600  3.000000e-02   
max      523.380000   2388.560000   8293.720000      8.584800  3.000000e-02   

                s17      s18      s19           s20           s21  
count  20631.000000  20631.0  20631.0  20631.000000  20631.000000  
mean     393.210654   2388.0    100.0     38.816271     23.289705  
std        1.548763      0.0      0.0      0.180746      0.108251  
min      388.000000   2388.0    100.0     38.140000     22.894200  
25%      392.000000   2388.0    100.0     38.700000     23.221800  
50%      393.000000   2388.0    100.0     38.830000     23.297900  
75%      394.000000   2388.0    100.0     38.950000     23.366800  
max      400.000000   2388.0    100.0     39.430000     23.618400  

[8 rows x 26 columns]

In this way, it's possible to check the fundamental statistics about each feature. 

We see that every feature has a count of 20631 values, which is the number of instances in the training set, in addition, it is straigthforward to investigate on their mean, standard deviation, minimum, maximum and quartiles values.

There is a total of 100 engines, each with a number of cycles between 1 and 362, with a mean of 108 cycles per engine; in this (training) dataset, the last cycle for each engine represents the last cycle of operation before failure.

Loading the test data:

In [7]:
df_test_raw = pd.read_csv('data/PM_test.txt', sep = ' ', header=None)
df_test_raw

0    1       2       3      4       5       6        7        8   \
0        1    1  0.0023  0.0003  100.0  518.67  643.02  1585.29  1398.21   
1        1    2 -0.0027 -0.0003  100.0  518.67  641.71  1588.45  1395.42   
2        1    3  0.0003  0.0001  100.0  518.67  642.46  1586.94  1401.34   
3        1    4  0.0042  0.0000  100.0  518.67  642.44  1584.12  1406.42   
4        1    5  0.0014  0.0000  100.0  518.67  642.51  1587.19  1401.92   
...    ...  ...     ...     ...    ...     ...     ...      ...      ...   
13091  100  194  0.0049  0.0000  100.0  518.67  643.24  1599.45  1415.79   
13092  100  195 -0.0011 -0.0001  100.0  518.67  643.22  1595.69  1422.05   
13093  100  196 -0.0006 -0.0003  100.0  518.67  643.44  1593.15  1406.82   
13094  100  197 -0.0038  0.0001  100.0  518.67  643.26  1594.99  1419.36   
13095  100  198  0.0013  0.0003  100.0  518.67  642.95  1601.62  1424.99   

          9   ...       18      19    20   21    22     23     24       25  \
0      14.62  ...  8125.55  8.4052  0.03  392  2388  100.0  38.86  23.3735   
1      14.62  ...  8139.62  8.3803  0.03  393  2388  100.0  39.02  23.3916   
2      14.62  ...  8130.10  8.4441  0.03  393  2388  100.0  39.08  23.4166   
3      14.62  ...  8132.90  8.3917  0.03  391  2388  100.0  39.00  23.3737   
4      14.62  ...  8129.54  8.4031  0.03  390  2388  100.0  38.99  23.4130   
...      ...  ...      ...     ...   ...  ...   ...    ...    ...      ...   
13091  14.62  ...  8213.28  8.4715  0.03  394  2388  100.0  38.65  23.1974   
13092  14.62  ...  8210.85  8.4512  0.03  395  2388  100.0  38.57  23.2771   
13093  14.62  ...  8217.24  8.4569  0.03  395  2388  100.0  38.62  23.2051   
13094  14.62  ...  8220.48  8.4711  0.03  395  2388  100.0  38.66  23.2699   
13095  14.62  ...  8214.64  8.4903  0.03  396  2388  100.0  38.70  23.1855   

       26  27  
0     NaN NaN  
1     NaN NaN  
2     NaN NaN  
3     NaN NaN  
4     NaN NaN  
...    ..  ..  
13091 NaN NaN  
13092 NaN NaN  
13093 NaN NaN  
13094 NaN NaN  
13095 NaN NaN  

[13096 rows x 28 columns]

The test data as the same original structure as the train one, so we can prepare it in the same manner as before:

In [8]:
df_test_raw.drop([26,27], axis=1, inplace=True)
df_test_raw.columns = col_names
df_test_raw

id  cycle  setting1  setting2  setting3      s1      s2       s3  \
0        1      1    0.0023    0.0003     100.0  518.67  643.02  1585.29   
1        1      2   -0.0027   -0.0003     100.0  518.67  641.71  1588.45   
2        1      3    0.0003    0.0001     100.0  518.67  642.46  1586.94   
3        1      4    0.0042    0.0000     100.0  518.67  642.44  1584.12   
4        1      5    0.0014    0.0000     100.0  518.67  642.51  1587.19   
...    ...    ...       ...       ...       ...     ...     ...      ...   
13091  100    194    0.0049    0.0000     100.0  518.67  643.24  1599.45   
13092  100    195   -0.0011   -0.0001     100.0  518.67  643.22  1595.69   
13093  100    196   -0.0006   -0.0003     100.0  518.67  643.44  1593.15   
13094  100    197   -0.0038    0.0001     100.0  518.67  643.26  1594.99   
13095  100    198    0.0013    0.0003     100.0  518.67  642.95  1601.62   

            s4     s5  ...     s12      s13      s14     s15   s16  s17   s18  \
0      1398.21  14.62  ...  521.72  2388.03  8125.55  8.4052  0.03  392  2388   
1      1395.42  14.62  ...  522.16  2388.06  8139.62  8.3803  0.03  393  2388   
2      1401.34  14.62  ...  521.97  2388.03  8130.10  8.4441  0.03  393  2388   
3      1406.42  14.62  ...  521.38  2388.05  8132.90  8.3917  0.03  391  2388   
4      1401.92  14.62  ...  522.15  2388.03  8129.54  8.4031  0.03  390  2388   
...        ...    ...  ...     ...      ...      ...     ...   ...  ...   ...   
13091  1415.79  14.62  ...  520.69  2388.00  8213.28  8.4715  0.03  394  2388   
13092  1422.05  14.62  ...  521.05  2388.09  8210.85  8.4512  0.03  395  2388   
13093  1406.82  14.62  ...  521.18  2388.04  8217.24  8.4569  0.03  395  2388   
13094  1419.36  14.62  ...  521.33  2388.08  8220.48  8.4711  0.03  395  2388   
13095  1424.99  14.62  ...  521.07  2388.05  8214.64  8.4903  0.03  396  2388   

         s19    s20      s21  
0      100.0  38.86  23.3735  
1      100.0  39.02  23.3916  
2      100.0  39.08  23.4166  
3      100.0  39.00  23.3737  
4      100.0  38.99  23.4130  
...      ...    ...      ...  
13091  100.0  38.65  23.1974  
13092  100.0  38.57  23.2771  
13093  100.0  38.62  23.2051  
13094  100.0  38.66  23.2699  
13095  100.0  38.70  23.1855  

[13096 rows x 26 columns]

We check again for `NaN` values:

In [9]:
df_test_raw.isnull().sum()

id          0
cycle       0
setting1    0
setting2    0
setting3    0
s1          0
s2          0
s3          0
s4          0
s5          0
s6          0
s7          0
s8          0
s9          0
s10         0
s11         0
s12         0
s13         0
s14         0
s15         0
s16         0
s17         0
s18         0
s19         0
s20         0
s21         0
dtype: int64

After we make sure that the test set is also good, we analyze it:

In [10]:
df_test_raw.describe()

id         cycle      setting1      setting2  setting3  \
count  13096.000000  13096.000000  13096.000000  13096.000000   13096.0   
mean      51.543907     76.836515     -0.000011      0.000004     100.0   
std       28.289423     53.057749      0.002203      0.000294       0.0   
min        1.000000      1.000000     -0.008200     -0.000600     100.0   
25%       28.000000     33.000000     -0.001500     -0.000200     100.0   
50%       52.000000     69.000000     -0.000000     -0.000000     100.0   
75%       76.000000    113.000000      0.001500      0.000300     100.0   
max      100.000000    303.000000      0.007800      0.000700     100.0   

             s1            s2            s3            s4            s5  ...  \
count  13096.00  13096.000000  13096.000000  13096.000000  1.309600e+04  ...   
mean     518.67    642.475088   1588.099204   1404.735362  1.462000e+01  ...   
std        0.00      0.400899      5.003274      6.688309  1.776425e-15  ...   
min      518.67    641.130000   1569.040000   1384.390000  1.462000e+01  ...   
25%      518.67    642.197500   1584.600000   1399.950000  1.462000e+01  ...   
50%      518.67    642.460000   1587.990000   1404.440000  1.462000e+01  ...   
75%      518.67    642.740000   1591.362500   1409.050000  1.462000e+01  ...   
max      518.67    644.300000   1607.550000   1433.360000  1.462000e+01  ...   

                s12           s13           s14           s15           s16  \
count  13096.000000  13096.000000  13096.000000  13096.000000  1.309600e+04   
mean     521.747724   2388.071025   8138.947817      8.425844  3.000000e-02   
std        0.559627      0.056934     10.188605      0.029009  1.040874e-17   
min      519.380000   2387.890000   8108.500000      8.332800  3.000000e-02   
25%      521.380000   2388.030000   8132.310000      8.405600  3.000000e-02   
50%      521.780000   2388.070000   8138.390000      8.424900  3.000000e-02   
75%      522.150000   2388.110000   8144.360000      8.444300  3.000000e-02   
max      523.760000   2388.320000   8220.480000      8.541400  3.000000e-02   

                s17      s18      s19           s20           s21  
count  13096.000000  13096.0  13096.0  13096.000000  13096.000000  
mean     392.571625   2388.0    100.0     38.892502     23.335743  
std        1.233577      0.0      0.0      0.141681      0.084120  
min      389.000000   2388.0    100.0     38.310000     22.935400  
25%      392.000000   2388.0    100.0     38.800000     23.281600  
50%      393.000000   2388.0    100.0     38.900000     23.339200  
75%      393.000000   2388.0    100.0     38.990000     23.393600  
max      397.000000   2388.0    100.0     39.410000     23.641900  

[8 rows x 26 columns]

### Test Dataset description
The ranges of values for the features in the test set are similar to the ones of the train, as we expected.

In this case there are roughly half as many instances (13096), and the cycles per engine (still 100 of them) range from 1 to 303 with an average of 76, though this time the last cycle for every engine doesn't represent the actual cycle before failure; that information is encoded in the last file of the three, the *truth* folder.

Loading the *truth* dataset:

In [14]:
df_truth = pd.read_csv('data/PM_truth.txt', sep = ' ', header=None)
df_truth

0   1
0   112 NaN
1    98 NaN
2    69 NaN
3    82 NaN
4    91 NaN
..  ...  ..
95  137 NaN
96   82 NaN
97   59 NaN
98  117 NaN
99   20 NaN

[100 rows x 2 columns]

As usual the structure of the dataframe is modified and cleaned:

In [15]:
df_truth.drop([1], axis=1, inplace=True)
df_truth.columns = ['ttf']
df_truth

ttf
0   112
1    98
2    69
3    82
4    91
..  ...
95  137
96   82
97   59
98  117
99   20

[100 rows x 1 columns]

### Truth Dataset description
The **time to failure** or **ttf** column shows how many cycles for each of the 100 engines in the test sets actually occurred before failure. Sometimes this feature is referred to as **Remaining Useful Life** or **RUL**, and is the target feature that the regression or classification model aims to predict.
For this reason the **ttf** feature must be extracted from the *truth* file and inserted as new feature in both the *train* and *test* datasets.

## Feature Extraction
In order to work efficiently, a predictive maintenance model needs to not only be able to analyze feature values of the current cycle, but also have an idea of what they where in previous cycles and how they changed in a defined precedent time window.

This concept is implemented through the creation of two extra features: called the **rolling average** and **rolling standard deviation**
1. **rolling average**: the average value of the feature calculated upon the last time window;
2. **rolling standard deviation**: the standard deviation of the feature calculated upon the last time window.

The function `add_features` does this job:

In [ ]:
def add_features(df_in, rolling_win_size):
    
    """Add rolling average and rolling standard deviation for sensors readings using fixed rolling window size.
    
    Args:
            df_in (dataframe)     : The input dataframe to be proccessed (training or test) 
            rolling_win_size (int): The window size, number of cycles for applying the rolling function
        
    Reurns:
            dataframe: contains the input dataframe with additional rolling mean and std for each sensor
    
    """
    
    sensor_cols = ['s1','s2','s3','s4','s5','s6','s7','s8','s9','s10','s11','s12','s13','s14','s15','s16','s17','s18','s19','s20','s21']
    sensor_av_cols = [nm.replace('s', 'av') for nm in sensor_cols]
    sensor_sd_cols = [nm.replace('s', 'sd') for nm in sensor_cols]
    df_out = pd.DataFrame()
    ws = rolling_win_size
    
    #calculate rolling stats for each engine id
    for m_id in pd.unique(df_in.id):
    
        # get a subset for each engine sensors
        df_engine = df_in[df_in['id'] == m_id]
        df_sub = df_engine[sensor_cols]

        # get rolling mean for the subset
        av = df_sub.rolling(ws, min_periods=1).mean()
        av.columns = sensor_av_cols
    
        # get the rolling standard deviation for the subset
        sd = df_sub.rolling(ws, min_periods=1).std().fillna(0)
        sd.columns = sensor_sd_cols
    
        # combine the two new subset dataframes columns to the engine subset
        new_ftrs = pd.concat([df_engine,av,sd], axis=1)
    
        # add the new features rows to the output dataframe
        df_out = pd.concat([df_out,new_ftrs])
        
    return df_out